Extract Story Drifts

GCRA101 opened the pull request:
### NOTE: Depends on 
https://github.com/BHoM/BHoM/pull/1639


### Issues addressed by this PR

 Closes #138 

![GH Script vs ETABS Tables](https://github.com/user-attachments/assets/24c5ba7a-130e-4a47-bd7d-74f3dc466738)

ETABS Toolkit now able to extract Story Drifts from ETABS Model based on input loadcase/loadcombo.

 ### Test files
Grasshopper File
https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23477-ExtractStoryDrifts/Test%20Script.gh?csf=1&web=1&e=lrVtVc
ETABS File
https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23477-ExtractStoryDrifts/Test%20ETABS%20Model.EDB?csf=1&web=1&e=Frkg7n


 ### Changelog

- Added StoryDrift Class to BHoM Structure Global Results
- Added GetStoryDrifts Method to ETABS Toolkit Read Global Results

peterjamesnugent commented on Etabs_Adapter/CRUD/Read/Results/GlobalResults.cs:
```diff
@@ -135,6 +138,39 @@ private List<ModalDynamics> GetModalParticipationMassRatios()
+            double[] x = null; double[] y= null; double[] z = null;
+
+            m_model.Results.StoryDrifts(ref resultCount, ref storyNames, ref loadcaseNames, ref stepType, ref stepNum, ref directions, ref drifts, ref labels, ref x, ref y, ref z);
+
+
+
```
remove whitespace

GCRA101 replied:
@peterjamesnugent white space removed ;)

peterjamesnugent commented on Etabs_Adapter/CRUD/Read/Results/GlobalResults.cs:
```diff
@@ -135,6 +138,39 @@ private List<ModalDynamics> GetModalParticipationMassRatios()
+                double timeStep;
+                GetStepAndMode(stepType[i], stepNum[i], out timeStep, out mode);
+
+                StoryDrift stDrft = new StoryDrift("", loadcaseNames[i], mode, timeStep, storyNames[i], directions[i], drifts[i]);
```
Just to confirm the `objectId` parameter is left blank because the results are global and not related to a specific object. The storey will specify the level.

```suggestion
                StoryDrift stDrft = new StoryDrift("", loadcaseNames[i], mode, timeStep, storeyNames[i], directions[i], drifts[i]);
```

GCRA101 replied:
@peterjamesnugent, changes incorporated

peterjamesnugent commented on Etabs_Adapter/CRUD/Read/Results/GlobalResults.cs:
```diff
@@ -135,6 +138,39 @@ private List<ModalDynamics> GetModalParticipationMassRatios()
+
+                StoryDrift stDrft = new StoryDrift("", loadcaseNames[i], mode, timeStep, storyNames[i], directions[i], drifts[i]);
+
+                storyDrifts.Add(stDrft);
```
```suggestion
                storeyDrifts.Add(stDrft);
```

GCRA101 replied:
@peterjamesnugent 
change incorporated in the code.

peterjamesnugent commented on Etabs_Adapter/CRUD/Read/Results/GlobalResults.cs:
```diff
@@ -135,6 +138,39 @@ private List<ModalDynamics> GetModalParticipationMassRatios()
+                double timeStep;
+                GetStepAndMode(stepType[i], stepNum[i], out timeStep, out mode);
+
+                StoryDrift stDrft = new StoryDrift("", loadcaseNames[i], mode, timeStep, storyNames[i], directions[i], drifts[i]);
```
```suggestion
                StoreyDrift stDrft = new StoreyDrift("", loadcaseNames[i], mode, timeStep, storeyNames[i], directions[i], drifts[i]);
```
This will need to reflected on the oM PR.

GCRA101 replied:
@peterjamesnugent 
change incorporated in the code.

peterjamesnugent commented on Etabs_Adapter/CRUD/Read/Results/GlobalResults.cs:
```diff
@@ -135,6 +138,39 @@ private List<ModalDynamics> GetModalParticipationMassRatios()
+            double[] drifts = null; string[] labels = null;
+            double[] x = null; double[] y= null; double[] z = null;
+
+            m_model.Results.StoryDrifts(ref resultCount, ref storyNames, ref loadcaseNames, ref stepType, ref stepNum, ref directions, ref drifts, ref labels, ref x, ref y, ref z);
```
```suggestion
            m_model.Results.StoreyDrifts(ref resultCount, ref storyNames, ref loadcaseNames, ref stepType, ref stepNum, ref directions, ref drifts, ref labels, ref x, ref y, ref z);
```

peterjamesnugent commented on Etabs_Adapter/CRUD/Read/Results/GlobalResults.cs:
```diff
@@ -56,6 +56,9 @@ public IEnumerable<IResult> ReadResults(GlobalResultRequest request, ActionConfi
                     return GetGlobalReactions();
                 case GlobalResultType.ModalDynamics:
+
+                case GlobalResultType.StoryDrifts:
```
```suggestion
                case GlobalResultType.StoreyDrifts:
```

GCRA101 replied:
@peterjamesnugent 
change incorporated in the code.

peterjamesnugent commented on Etabs_Adapter/CRUD/Read/Results/GlobalResults.cs:
```diff
@@ -56,6 +56,9 @@ public IEnumerable<IResult> ReadResults(GlobalResultRequest request, ActionConfi
                 case GlobalResultType.ModalDynamics:
+
+                case GlobalResultType.StoryDrifts:
+                    return GetStoryDrifts();
```
```suggestion
                    return GetStoreyDrifts();
```

GCRA101 replied:
@peterjamesnugent 
change incorporated in the code.

peterjamesnugent reviewed (changes requested):
Some changes relating to the naming - can probably just commit those and get someone to review functionality.

HugoVanLooveren commented:
I tried building the solution and it gave an error:
Tried with Beta 7.2 version.
Also tried with Alpha 241024 version.
![image](https://github.com/user-attachments/assets/612e678d-f3ae-4f1c-afdb-b0cd2a9d1ac1)

peterjamesnugent commented:
@HugoVanLooveren you need to build the branch from the core BHoM as per the PR notes:
https://github.com/BHoM/BHoM/pull/1639

HugoVanLooveren reviewed (changes requested):
1. Issue with building it because of missing enum value for GlobalResultType.cs, see comment Peter in [#1639](https://github.com/BHoM/BHoM/pull/1639).

2. Naming convention singular and plural seems inconsistent between this PR and #1639. 
Line 60 & 61 particularly. (Easiest seems to change #1639 and the enum file to be plural, which than is consistent with GlobalResultType.Reactions & GlobalResultType.ModalDynamics)

3. After a manual change in the enum and line 60 (case GlobalResultType.StoreyDrifts minus the s) the build worked. Using ETABS 21 and running the test files resulted in the same results in GH as in ETABS table.

GCRA101 commented:
@peterjamesnugent , @HugoVanLooveren ,

updated the code in line with changes to BHoM Toolkit and comments above.
Rebuilt it and works ok ;)
Updated code in the latest commit pushed just now ;)

![image](https://github.com/user-attachments/assets/742c066d-0006-41fe-86fe-fe02c1fd3274)

HugoVanLooveren reviewed (approved):
Reviewed again after last commit with ETABS 21.
Works as would be expected, so approved from my side :)

peterjamesnugent commented:
@BHoMBot check required

peterjamesnugent commented:
@BHoMBot check copyright-compliance

peterjamesnugent commented:
@BHoMBot check core

GCRA101 commented:
@BHoMBot check required

GCRA101 commented:
@peterjamesnugent , @HugoVanLooveren ,
updated the code turning the Direction attribute from string to vector (see latest commit 96184c7d1b9f443728e3d98bc8072a1dd399678a).

Tested against Test GH Script (see it at link https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23477-ExtractStoryDrifts/Test%20Script.gh?csf=1&web=1&e=lrVtVc)
and all works fine (see image below)

![StoryDrifts Outputs](https://github.com/user-attachments/assets/ad79127b-e3de-4c5a-a7ac-89c0081ddb7c)

GCRA101 commented:
@BHoMBot check required

GCRA101 commented:
@BHoMBot check copyright-compliance

GCRA101 commented:
@BHoMBot ready-to-merge

GCRA101 commented:
@BHoMBot check ready-to-merge

GCRA101 commented on Etabs_Adapter/CRUD/Read/Results/GlobalResults.cs:
```diff
@@ -135,6 +139,54 @@ private List<ModalDynamics> GetModalParticipationMassRatios()
+
+            m_model.Results.StoryDrifts(ref resultCount, ref storeyNames, ref loadcaseNames, ref stepType, ref stepNum, ref directions, ref drifts, ref labels, ref x, ref y, ref z);
+
+            Vector[] dirVectors = directions.Select(dirStr =>
+                {
+                    switch (dirStr.ToUpper())
+                    {
+                        case "X":
+                            return Vector.XAxis;
+                            break;
+                        case "Y":
+                            return Vector.YAxis;
+                            break;
+                        default:
+                            return Vector.ZAxis;
+                            break;
+                    }
+                }).ToArray();
```
@peterjamesnugent, thanks for the comment.
I think that this can stay. The directions output from ETABS via its API (in the code line above the snippet), are string objects with value "X"/"Y"/"Z".
The switch case block above is converting the string data into BHoM Vector data to be assigned to the BHoM StoreyDrift object.
Let me know in case I'm missing sth ;)

peterjamesnugent commented:
@BHoMBot check ready-to-merge

GCRA101 commented:
@BHoMBot check ready-to-merge

pawelbaran commented:
@BHoMBot check ready-to-merge

GCRA101 commented:
@BHoMBot check ready-to-merge

HugoVanLooveren reviewed (approved):
Tested functionality with BHoM Beta 7.2.1.0
Tested with ETABS 21

Works as expected, numbers were matching. All good :)

peterjamesnugent commented:
@BHoMBot check ready-to-merge
